src/Capability/Registry.php

```diff
@@ -238,30 +241,173 @@ public function getResourceTemplate(string $uriTemplate): ?ResourceTemplateRefer
+    {
+        $prompts = [];
+        foreach ($this->prompts as $promptReference) {
+            $prompts[] = $promptReference->prompt;
+        }
+
+        if (null === $limit) {
+            return $prompts;
+        }
+
+        return $this->paginateResults($prompts, $limit, $cursor);
     }
 
-    public function getResources(): array
+    /**
+     *  Gets all registered resource templates.
+     *
+     * @return list<ResourceTemplate>
+     */
+    public function getResourceTemplates(?int $limit = null, ?string $cursor = null): array
     {
-        return array_map(fn (ResourceReference $resource) => $resource->schema, $this->resources);
+        $templates = [];
+        foreach ($this->resourceTemplates as $templateReference) {
+            $templates[] = $templateReference->resourceTemplate;
+        }
+
+        if (null === $limit) {
+            return $templates;
+        }
+
+        return $this->paginateResults($templates, $limit, $cursor);
     }
 
-    public function getPrompts(): array
+    /**
+     * Helper method to paginate results using cursor-based pagination.
+     *
+     * @param list<mixed> $items  The full array of items to paginate
+     * @param int         $limit  Maximum number of items to return
+     * @param string|null $cursor Base64 encoded offset position
+     *
+     * @return list<mixed> Paginated results
+     *
+     * @throws InvalidCursorException When cursor is invalid (MCP error code -32602)
+     */
+    private function paginateResults(array $items, int $limit, ?string $cursor = null): array
     {
-        return array_map(fn (PromptReference $prompt) => $prompt->prompt, $this->prompts);
+        $offset = 0;
+        if (null !== $cursor) {
+            $decodedCursor = base64_decode($cursor, true);
+
+            if (false === $decodedCursor || !is_numeric($decodedCursor)) {
+                throw new InvalidCursorException($cursor);
+            }
+
+            $offset = (int) $decodedCursor;
+
+            // Validate offset is within reasonable bounds
+            if ($offset < 0 || $offset > \count($items)) {
+                throw new InvalidCursorException($cursor);
+            }
+        }
+
+        $result = [];
+        $count = 0;
+        $currentIndex = 0;
+
+        foreach ($items as $item) {
+            if ($currentIndex >= $offset && $count < $limit) {
+                $result[] = $item;
+                ++$count;
+            }
+            ++$currentIndex;
+        }
+
+        return $result;
     }
 
-    public function getResourceTemplates(): array
+    /**
+     * Calculate next cursor for pagination.
+     *
+     * @param list<mixed> $allItems      The complete array of items
+     * @param string|null $currentCursor Current cursor position
+     * @param int         $returnedCount Number of items actually returned
+     */
+    public function calculateNextCursor(array $allItems, ?string $currentCursor, int $returnedCount): ?string
```

Comment:
if we're only doing a `count()` on `$allItems` it should be the count as `int` parameter right away

Reply:
Ah ! I see what you're saying. Will just count within the getTools and parse. Make sense